Core/APIRequest.swift

```diff
@@ -21,26 +21,87 @@
 import Foundation
 import Alamofire
 
-public typealias APIRequestCompletion = (Data?, Error?) -> Swift.Void
+public enum APIRequestCompleteionResult {
+
+    case dataPersisted
+    case errorHandled
+
+}
+
+public typealias APIRequestCompletion = (Data?, Error?) -> APIRequestCompleteionResult
 
 public class APIRequest {
 
+    struct HeaderNames {
+
+        static let etag = "ETag"
+
+    }
+
     let url: URL
+    var etagStorage: APIRequestETagStorage
 
-    init(url: URL) {
+    init(url: URL, etagStorage: APIRequestETagStorage = UserDefaultsETagStorage()) {
         self.url = url
+        self.etagStorage = etagStorage
     }
 
     func execute(completion: @escaping APIRequestCompletion) {
+        let etag = etagStorage.etag(for: url)
+
+        Logger.log(text: "Requesting \(url)")
 
-        Logger.log(text: "Requesting \(url) ...")
         Alamofire.request(url)
             .validate(statusCode: 200..<300)
             .responseData(queue: DispatchQueue.global(qos: .utility)) { response in
-                Logger.log(text: "Request for \(self.url) completed with result \(response.result)")
-                completion(response.data, response.error)
-        }
+
+                Logger.log(text: "Request for \(self.url) completed with response code: \(String(describing: response.response?.statusCode)) and headers \(String(describing: response.response?.allHeaderFields))")
+
+                if etag != nil && etag == response.response?.headerValue(for: HeaderNames.etag) {
+                    Logger.log(text: "Using cached version of \(self.url) with etag: \(String(describing: etag))")
+                    _ = completion(nil, nil)
```

Comment:
Does returning `nil` instead of data here essentially break the cache? I think returning an optional etag in the completion would work better than doing all this magic here and suppressing the data. Then the caller can store the etag if they care about it and choose how to proceed. It also avoids brittle url / etag lookups.

Reply:
There’s no magic here. As an API request it seems reasonable for this class to handle any caching. The caller doesn’t need to know about that and should just handle data if there is some. But the completion handler must be called so nils are appropriate as that’s the same as saying “no new data” and “no error”. Ie the lack of error does not imply the presence of data.

As long as the caller doesn’t persist nil (why would it?) it should be fine.

Are you seeing a problem running the code?

Reply:
I think the problem might be that this is an API” request” but can handle any URL. It might be better if this was more explicitly a content blocker list request which takes an enum case to build the URL and use that as the key for the etag cache.

I could even add that as an extra layer of abstraction between APIRequest and the BlockerListDownloader but my preference would just be to rename it and make the change as we don’t have a strong requirement for a generic downloading interface that I’m aware of and YAGNI.




I’ll uodate the code to be more specific.

Reply:
Actually, I can see you're using it - I'll add the abstraction. :)

Reply:
Sounds good Chris!